### ISSUE TITLE
System.Xml.XmlDocument.Save() unexpectedly creates UTF-8 files *with BOM*

### ISSUE THREAD

**Issue by mklement0** [dotnet/corefx#34118]
`System.Xml.XmlDocument.Save()`, when given a file path to write to, unexpectedly creates UTF-8 encoded files _with BOM_ if the document's XML declaration has an `encoding="UTF-8"` attribute.

(By contrast, the absence of the `encoding` attribute or absence of an XML declaration altogether causes a BOM-_less_ UTF-8 file to be created, as expected.)

Repro in PowerShell Core (on any supported platform; PowerShell's `[xml]` cast instantiates a `System.Xml.XmlDocument` from the specified XML string):

```powershell
([xml] '<?xml version="1.0" encoding="UTF-8" ?><foo />').Save("$HOME/t.xml")
Get-Content -AsByteStream $HOME/t.xml | Select-Object -First 3
```

The above yields:

```none
239
187
191
```

which are the byte values of the UTF-8 BOM (`0xef 0xbb 0xbf`)

**Comment by danmosemsft** [comment]
Whta does it do in .NET Framework?

**Comment by mklement0** [comment]
@danmosemsft: 

It does the **same as in .NET Core;** you can verify that by running the equivalent command in _Windows PowerShell_ (which requires substitution `-Encoding Byte` for `-AsByteStream`):

```powershell
([xml] '<?xml version="1.0" encoding="UTF-8" ?><foo />').Save("$HOME/t.xml")
# Same output as in .NET Core.
Get-Content -Encoding Byte $HOME/t.xml | Select-Object -First 3
```

**which it should never have done**, given that .NET's (sensible) default since v1 has  been BOM-_less_ UTF-8.

**Comment by danmosemsft** [comment]
Question is how breaking it would be to change.

**Comment by mklement0** [comment]
@danmosemsft:

In the abstract:

Any code that later consumes `.Save()`-created files and assumes that the _absence_ of a BOM implies something _other_ than UTF-8 is hypothetically affected.

.NET code that doesn't specify an encoding explicitly when reading such files is _not_ affected.

I can't tell you what that means in terms of real-world impact.

If, sadly, it turns out that changing this would be too much of a breaking change, this historical wart _must be documented_.

**Comment by mklement0** [comment]
@danmosemsft: 

Re real-world impact with respect to _PowerShell_: 

PowerShell _Core_ now defaults to UTF-8 when reading a file without a BOM, so it wouldn't be affected.

Windows PowerShell defaults to the legacy "ANSI" code page, so code reading a BOM-less UTF-8 file _as a text file_ with `Get-Content` without also specifying `-Encoding Utf8` _is_ affected; however, loading the file via `($xml = New-Object System.Xml.XmlDocument).Load("/path/to/file.xml")`is again not affected.

**Comment by danmosemsft** [comment]
It looks like it was this way in v2.0 of .NET Framework. I guess XML area owners should make a call on this one.

**Comment by krwq** [comment]
I do not know the Powershell syntax for this but from Xml perspective this behavior looks correct and expected. If you are able to specify the encoding as: `new UTF8Encoding(false)` (see: https://docs.microsoft.com/en-us/dotnet/api/system.text.utf8encoding.-ctor?view=netframework-4.7.2#System_Text_UTF8Encoding__ctor_System_Boolean_ for more details) then I believe the BOM should disappear. Here is the example in C#:

```csharp
    XmlDocument doc = ...; // construct your doc here
    XmlWriterSettings settings = new XmlWriterSettings();
    settings.Encoding = new UTF8Encoding(false);
    XmlWriter writer = XmlWriter.Create(stream, settings);
    doc.Save(writer);
```

I'm closing this as by design, if you can still see BOM after setting the encoding like this please reopen.

**Comment by mklement0** [comment]
@krwq: 

PowerShell is incidental to this issue; I merely used it to demonstrate the problem succinctly.

To demonstrate the issue using C#:

```csharp
using System.IO;

var xmlDom = new System.Xml.XmlDocument();
xmlDom.LoadXml("<?xml version=\"1.0\" encoding=\"UTF-8\" ?><foo>bär</foo>");
xmlDom.Save($@"{Environment.GetFolderPath(Environment.SpecialFolder.UserProfile)}/t.xml");
```

If you inspect the resulting `t.xml` file in the user's home folder, you'll see that it - unexpectedly - has a UTF-8 "BOM" (Unicode signature).

That you can _work around_ the issue by using an explicit BOM-less encoding, as in your code, is beside the point.

The point is that .NET, since v1, has defaulted to BOM-less UTF-8.

_Explicitly_ opting to use UTF-8 _as part of the XML declaration_ shouldn't change that behavior, yet it currently does: if you omit the `encoding="UTF-8"` part of the declaration, you _do_ get a BOM-less UTF-8 file. There is no reason to act differently with `encoding="UTF-8"` present, and doing so hurts portability.

Therefore, please reopen this issue.

**Comment by krwq** [comment]
@mklement0 I normally would expect BOM as it leaves less fuzziness about what user have meant. Have you tried saving XmlDocument the way I've sent you above?

**Comment by krwq** [comment]
@mklement0 if you do not specify declaration then you are not intentional in how do you want the xml writer to behave.
Long time ago we chose that the default will be utf-8 without BOM. Since we have behaved like that for a very long time it would be very breaking to change it right now.

If you use encoding in the declaration then you are intentional and asking XML writer to write utf-8. The most portable way to write text to a file is to add BOM.

Imagine you write declaration which suggests UTF-16 encoding. Should we write BOM now or not? Should we specially treat UTF-8? If you don't write encoding you maybe have meant ASCII with some old code page? Note that the declaration itself is also encoded in the encoding it talks about so parser have to have some idea about encoding earlier so writing BOM is the right thing.

**Comment by mklement0** [comment]
@krwq:

> Long time ago we chose that the default will be utf-8 without BOM.

Indeed. Thank {supreme being of your choice}. There is absolutely no reason to change that, and it certainly isn't what this issue is proposing.

> The most portable way to write text to a file is to add BOM.

**Not at all: on Unix-like platforms, native utilities neither expect nor know how to handle a UTF-8 "BOM"** - they treat the BOM as _data_, which obviously causes problems: see https://en.wikipedia.org/wiki/UTF-8#Byte_order_mark

.NET wisely chose UTF-8 as the _default_, in the _absence_ of a BOM (Unicode signature); in effect, it is what the Unix world has done as well.

To summarize:

* From a cross-platform perspective, writing a "BOM" for UTF-8-encoded data should be avoided.

* From an internal-consistency perspective, exhibiting different behavior if what is the implied default is specified explicitly should be avoided.

Therefore, please reopen this issue.

**Comment by mklement0** [comment]
@krwq: Please don't ignore this issue: **Writing  UTF-8 files _with BOM_ is a serious impediment to portability and at odds with .NET's default**.
In the Windows-only days this may have been less of a concern, but in .NET Core it matters.

There is no reason for the implied and explicit use of UTF-8 to act differently; specifically, there should be no difference in behavior between these two declarations, and they both should result in BOM-_less_ UTF-8.

```
<?xml version="1.0"?>
<?xml version="1.0" encoding="UTF-8"?>
```

UTF-8 is the effective _default_ character encoding for XML, and 
while the [W3C XML Recommendation _allows_ a BOM with UTF-8](https://www.w3.org/TR/xml/#charencoding) (and every compliant XML parser should therefore support it):

* a UTF-8 BOM is never _needed_.

* there a good reasons _not_ to write one (as stated):
   * Portability: Unix[-heritage] utilities won't process the BOM correctly - note that while, ideally, XML files should always be handled by an XML _parser_, in practice they are often read as _text files_.
  * Internal consistency: Merely making what is the default encoding explicit should not change the behavior.

**Comment by krwq** [comment]
@mklement0 I understand your frustration here but this is one of the changes which is fairly risky: there is very little to gain by changing this - if we shipped 1st version of .NET now I'd probably use UTF-8 without BOM as well - but we have already shipped like that for many years - people start moving from full framework and this would be another difference in behavior which is not needed and might possibly be a blocker for someone. I've seen tinier changes to break people than that - the most stupid example is that hash code will be different. The worst case scenario is that some other tool has bad heuristics for detecting encoding and causing xml parser to think the file is incorrect.

BOM is needed for parsers which support encodings such as UTF-16 so that implementation can distinguish between little and big endian, UTF-8 or other encoding without guessing. Linux is UTF-8 oriented so UTF-16 or other encoding do not matter in that world (and that is really good and I like it as well) and arguably we should not care about it. In .NET world both UTF-8 and UTF-16 matter equally (especially string is UTF-16 by default which even maybe puts us slightly toward UTF-16) - we really care about back compat here and considering as you said w3c recommendation is that every xml parser should support BOM so there is simply no reason for us to risk breaking someone's app.

I.e.: `<` is 3C in hex and same for UTF-8 representation. For UTF-16LE the hex representation is `3C 00` so now you can imagine that parser sees `<` after which it may assume i.e. ASCII/UTF8, then instead of expected `?` it sees null character and reports error while previously it saw BOM and detected encoding correctly.

**Comment by mklement0** [comment]
I've created a docs issue: https://github.com/dotnet/docs/issues/10326

**Comment by krwq** [comment]
Thanks @mklement0
